src/box/lua/config/configdata.lua

```diff
@@ -530,7 +606,109 @@ local function new(iconfig, cconfig, instance_name)
                 found.replicaset_name), 0)
         end
         assert(bootstrap_leader == nil)
-        bootstrap_leader_name = peer_names[1]
+
+        -- Choose first non-anonymous instance.
+        for _, peer_name in ipairs(peer_names) do
+            assert(peers[peer_name] ~= nil)
+            local iconfig_def = peers[peer_name].iconfig_def
+            local is_anon = instance_config:get(iconfig_def, 'replication.anon')
+            if not is_anon then
+                bootstrap_leader_name = peer_name
+                break
+            end
+        end
+        assert(bootstrap_leader_name ~= nil)
+    end
+
+    -- Verify replication.anon = true prerequisites.
+    --
+    -- The key idea of all the checks are that an anonymous
+    -- replica must be in the read-only mode.
+    --
+    -- Different failover modes control read-only/read-write mode
+    -- in different ways, so we need specific checks for each of
+    -- them in regard of an anonymous replica.
+    --
+    -- These checks don't verify the whole cluster config, only
+    -- the given replicaset.
+
+    -- failover: <any>
+    --
+    -- A replicaset can't consist of only anonymous replicas.
+    assert(#peer_names > 0)
```

Comment:
Sure.

Also, moved it before `supervised` mode logic, which chooses a bootstrap leader, because it assumes that there is at least one non-anonymous instance in the given replicaset.